improve karmadactl unit test: check the cert file exist

tiansuo114 opened the pull request:
**What type of PR is this?**

<!--
Add one of the following kinds:
/kind feature
-->

**What this PR does / why we need it**:
[issue:5171](https://github.com/karmada-io/karmada/issues/5171)
**Which issue(s) this PR fixes**:
Fixes # [issue:5171](https://github.com/karmada-io/karmada/issues/5171)

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note

```

liangyuanpeng commented:
please following the error message to fix the lint and you may need to update the pr body `Fixes # https://github.com/karmada-io/karmada/issues/5171` to `Fixes #5171`

```shell
Run hack/verify-staticcheck.sh
Installing golangci-lint v1.59.0
golangci/golangci-lint info checking GitHub for tag 'v1.59.0'
golangci/golangci-lint info found version: 1.59.0 for v1.59.0/linux/amd6[4](https://github.com/karmada-io/karmada/actions/runs/9885709690/job/27304218303?pr=5175#step:6:5)
golangci/golangci-lint info installed /home/runner/go/bin/golangci-lint
pkg/karmadactl/cmdinit/cert/cert_test.go:21: File is not `gci`-ed with --skip-generated -s Standard -s Default -s Prefix(github.com/karmada-io/karmada) (gci)
	"k8s.io/klog/v2"
pkg/karmadactl/cmdinit/cert/cert_test.go:27: File is not `gci`-ed with --skip-generated -s Standard -s Default -s Prefix(github.com/karmada-io/karmada) (gci)
	certutil "k8s.io/client-go/util/cert"
pkg/karmadactl/cmdinit/cert/cert_test.go:2[6](https://github.com/karmada-io/karmada/actions/runs/9885709690/job/27304218303?pr=5175#step:6:7): File is not `goimports`-ed with -local github.com/karmada-io/karmada (goimports)


Please review the above warnings.
Tips: The golangci-lint might help you fix some issues, try with the command "golangci-lint run --fix".
If the above warnings do not make sense, feel free to file an issue.
```

liangyuanpeng commented:
may better title: `improve karmadactl unit test: check the cert file exist `

tiansuo114 commented:
> please following the error message to fix the lint and you may need to update the pr body `Fixes # https://github.com/karmada-io/karmada/issues/5171` to `Fixes #5171`
> 
> ```shell
> Run hack/verify-staticcheck.sh
> Installing golangci-lint v1.59.0
> golangci/golangci-lint info checking GitHub for tag 'v1.59.0'
> golangci/golangci-lint info found version: 1.59.0 for v1.59.0/linux/amd6[4](https://github.com/karmada-io/karmada/actions/runs/9885709690/job/27304218303?pr=5175#step:6:5)
> golangci/golangci-lint info installed /home/runner/go/bin/golangci-lint
> pkg/karmadactl/cmdinit/cert/cert_test.go:21: File is not `gci`-ed with --skip-generated -s Standard -s Default -s Prefix(github.com/karmada-io/karmada) (gci)
> 	"k8s.io/klog/v2"
> pkg/karmadactl/cmdinit/cert/cert_test.go:27: File is not `gci`-ed with --skip-generated -s Standard -s Default -s Prefix(github.com/karmada-io/karmada) (gci)
> 	certutil "k8s.io/client-go/util/cert"
> pkg/karmadactl/cmdinit/cert/cert_test.go:2[6](https://github.com/karmada-io/karmada/actions/runs/9885709690/job/27304218303?pr=5175#step:6:7): File is not `goimports`-ed with -local github.com/karmada-io/karmada (goimports)
> 
> 
> Please review the above warnings.
> Tips: The golangci-lint might help you fix some issues, try with the command "golangci-lint run --fix".
> If the above warnings do not make sense, feel free to file an issue.
> ```

tiansuo114 commented:
> > please following the error message to fix the lint and you may need to update the pr body `Fixes # https://github.com/karmada-io/karmada/issues/5171` to `Fixes #5171`
> > ```shell
> > Run hack/verify-staticcheck.sh
> > Installing golangci-lint v1.59.0
> > golangci/golangci-lint info checking GitHub for tag 'v1.59.0'
> > golangci/golangci-lint info found version: 1.59.0 for v1.59.0/linux/amd6[4](https://github.com/karmada-io/karmada/actions/runs/9885709690/job/27304218303?pr=5175#step:6:5)
> > golangci/golangci-lint info installed /home/runner/go/bin/golangci-lint
> > pkg/karmadactl/cmdinit/cert/cert_test.go:21: File is not `gci`-ed with --skip-generated -s Standard -s Default -s Prefix(github.com/karmada-io/karmada) (gci)
> > 	"k8s.io/klog/v2"
> > pkg/karmadactl/cmdinit/cert/cert_test.go:27: File is not `gci`-ed with --skip-generated -s Standard -s Default -s Prefix(github.com/karmada-io/karmada) (gci)
> > 	certutil "k8s.io/client-go/util/cert"
> > pkg/karmadactl/cmdinit/cert/cert_test.go:2[6](https://github.com/karmada-io/karmada/actions/runs/9885709690/job/27304218303?pr=5175#step:6:7): File is not `goimports`-ed with -local github.com/karmada-io/karmada (goimports)
> > 
> > 
> > Please review the above warnings.
> > Tips: The golangci-lint might help you fix some issues, try with the command "golangci-lint run --fix".
> > If the above warnings do not make sense, feel free to file an issue.
> > ```
> 
> I have rewritten the code and tested it locally using golangci-lint. It seems to have passed most of the tests, but I encountered a new issue in the e2e tests.
> 
> I noticed that all the e2e tests passed in the last commit, but after merging the two commits using git rebase without changing the code, I got an error in the CI. I don't quite understand the cause of the error.
> 
> Additionally, I would like to know what subcommands are used when running golangci-lint in GitHub CI or where I can see the hack/verify-staticcheck.sh file. When I run golangci-lint run locally, I see some other issues, and using --fix on this basis does not resolve the issues but rather causes more problems, which is quite strange.



> > please following the error message to fix the lint and you may need to update the pr body `Fixes # https://github.com/karmada-io/karmada/issues/5171` to `Fixes #5171`
> > ```shell
> > Run hack/verify-staticcheck.sh
> > Installing golangci-lint v1.59.0
> > golangci/golangci-lint info checking GitHub for tag 'v1.59.0'
> > golangci/golangci-lint info found version: 1.59.0 for v1.59.0/linux/amd6[4](https://github.com/karmada-io/karmada/actions/runs/9885709690/job/27304218303?pr=5175#step:6:5)
> > golangci/golangci-lint info installed /home/runner/go/bin/golangci-lint
> > pkg/karmadactl/cmdinit/cert/cert_test.go:21: File is not `gci`-ed with --skip-generated -s Standard -s Default -s Prefix(github.com/karmada-io/karmada) (gci)
> > 	"k8s.io/klog/v2"
> > pkg/karmadactl/cmdinit/cert/cert_test.go:27: File is not `gci`-ed with --skip-generated -s Standard -s Default -s Prefix(github.com/karmada-io/karmada) (gci)
> > 	certutil "k8s.io/client-go/util/cert"
> > pkg/karmadactl/cmdinit/cert/cert_test.go:2[6](https://github.com/karmada-io/karmada/actions/runs/9885709690/job/27304218303?pr=5175#step:6:7): File is not `goimports`-ed with -local github.com/karmada-io/karmada (goimports)
> > 
> > 
> > Please review the above warnings.
> > Tips: The golangci-lint might help you fix some issues, try with the command "golangci-lint run --fix".
> > If the above warnings do not make sense, feel free to file an issue.
> > ```
> 
> I have rewritten the code and tested it locally using golangci-lint. It seems to have passed most of the tests, but I encountered a new issue in the e2e tests.
> 
> I noticed that all the e2e tests passed in the last commit, but after merging the two commits using git rebase without changing the code, I got an error in the CI. I don't quite understand the cause of the error.
> 
> Additionally, I would like to know what subcommands are used when running golangci-lint in GitHub CI or where I can see the hack/verify-staticcheck.sh file. When I run golangci-lint run locally, I see some other issues, and using --fix on this basis does not resolve the issues but rather causes more problems, which is quite strange.

Now all the check items pass the status again. I solved the error by re-performing the ci process without changing the code, which seems a little strange

liangyuanpeng commented:
> where I can see the hack/verify-staticcheck.sh file.

https://github.com/karmada-io/karmada/blob/master/hack/verify-staticcheck.sh

>  I solved the error by re-performing the ci process without changing the code, which seems a little strange

CI is green and that's good :)

tiansuo114 commented:
> > where I can see the hack/verify-staticcheck.sh file.
> 
> https://github.com/karmada-io/karmada/blob/master/hack/verify-staticcheck.sh
> 
> > I solved the error by re-performing the ci process without changing the code, which seems a little strange
> 
> CI is green and that's good :)

This is good, but I see that there are still two checkpoints in the failed state, what else should I do to get them through

liangyuanpeng commented:
> This is good, but I see that there are still two checkpoints in the failed state, what else should I do to get them through

All CI is green now, and the yellow point is fine.

let we waiting for https://github.com/karmada-io/karmada/pull/5127 is merged and then update the test to cover the two new parameters,Thanks for your work!

liangyuanpeng commented on pkg/karmadactl/cmdinit/cert/cert_test.go:
```diff
@@ -104,4 +105,30 @@ func TestGenCerts(_ *testing.T) {
 	if err := GenCerts(TestCertsTmp, etcdServerCertConfig, etcdClientCertCfg, karmadaCertCfg, apiserverCertCfg, frontProxyClientCertCfg); err != nil {
 		fmt.Println(err)
```
Could you please help to update it to `t.Fatal(err)` and all of samhe problem from this test.

tiansuo114 replied:
Ok, I will replace all fmt.Println(err) with t.ror for the file concerned, but I don't quite understand what the samhe problem means, I tried to search through google, but I didn't get some possibly relevant answers

liangyuanpeng replied:
@tiansuo114 
sorry for typo,` samhe problem` --> `same problem`

liangyuanpeng reviewed (commented):
https://github.com/karmada-io/karmada/pull/5127 is merged and please go ahead!

liangyuanpeng commented on pkg/karmadactl/cmdinit/cert/cert_test.go:
```diff
@@ -107,6 +108,32 @@ func TestGenCerts(t *testing.T) {
+		"front-proxy-client.crt", "front-proxy-client.key",
+		"karmada.crt", "karmada.key",
+	}
+	if err := checkCertFiles(TestCertsTmp, certFiles); err != nil {
```
if TestCaCertPath and TestCaKeyPath is not nil and i think this is not TestCertsTmp folder anymore.

tiansuo114 replied:
I don't quite understand the amendment here, could you give me clearer instructions

liangyuanpeng replied:
same with https://github.com/karmada-io/karmada/pull/5175#discussion_r1680374431

we need to check two case:

1. TestCaCertPath and TestCaKeyPath  is nil
2. TestCaCertPath and TestCaKeyPath  is not nil

tiansuo114 replied:
I see, I will change it in the latest commit

liangyuanpeng reviewed (commented):
/cc @zhzhuang-zju  @chaosi-zju 
for review

chaosi-zju commented on pkg/karmadactl/cmdinit/cert/cert_test.go:
```diff
@@ -107,6 +108,32 @@ func TestGenCerts(t *testing.T) {
+
+func checkCertFiles(path string, files []string) error {
+	for _, file := range files {
+		filePath := fmt.Sprintf("%s/%s", path, file)
```
I would prefer like this:

```suggestion
		filePath := filepath.Join(path, file)
```

tiansuo114 replied:
I have modified this code in the latest rebase, please check it again

chaosi-zju commented on pkg/karmadactl/cmdinit/cert/cert_test.go:
```diff
@@ -107,6 +108,32 @@ func TestGenCerts(t *testing.T) {
+func checkCertFiles(path string, files []string) error {
+	for _, file := range files {
+		filePath := fmt.Sprintf("%s/%s", path, file)
+		if _, err := os.Stat(filePath); os.IsNotExist(err) {
```
how about other type error?

liangyuanpeng replied:
yes, we need to handle all of error.

tiansuo114 replied:
I have modified this code in the latest rebase, please check it again

chaosi-zju commented on pkg/karmadactl/cmdinit/cert/cert_test.go:
```diff
@@ -107,6 +108,32 @@ func TestGenCerts(t *testing.T) {
 	}
 	if err := GenCerts(TestCertsTmp, TestCaCertPath, TestCaKeyPath, etcdServerCertConfig, etcdClientCertCfg, karmadaCertCfg, apiserverCertCfg, frontProxyClientCertCfg); err != nil {
-		fmt.Println(err)
+		t.Fatal(err)
```
we call `GenCerts` twice by different internal implementation, since you want to `checkCertFiles`, it is best to test fully and test both.

chaosi-zju commented on pkg/karmadactl/cmdinit/cert/cert_test.go:
```diff
@@ -107,6 +108,32 @@ func TestGenCerts(t *testing.T) {
-		fmt.Println(err)
+		t.Fatal(err)
+	}
+
```
we call GenCerts twice by different internal implementation, since you want to checkCertFiles, it is best to test fully and test both.

tiansuo114 replied:
I have modified this code in the latest rebase, please check it again

chaosi-zju replied:
thanks~ But this is not perfect enough, supposing first `GenCerts` success and second `GenCerts` fails, since the certs generated by first `GenCerts` still exist, it can be detected by second `checkCertFiles`.

maybe we can set the output dir of second `GenCerts` to other path?

zhzhuang-zju commented on pkg/karmadactl/cmdinit/cert/cert_test.go:
```diff
@@ -71,7 +84,7 @@ func TestGenCerts(t *testing.T) {
 	}
 	if hostName, err := os.Hostname(); err != nil {
-		fmt.Println("Failed to get the current hostname.", err)
+		t.Fatal("Failed to get the current hostname.", err)
```
Actually，`internetIP, err := utils.InternetIP()` and `internetIP, err := utils.InternetIP()`  just generating some non-essential materials, and is not related to the success of `GenCerts`. So, I think there is no need to use `t. Fatal`.

tiansuo114 replied:
> Actually，`internetIP, err := utils.InternetIP()` and `internetIP, err := utils.InternetIP()` just generating some non-essential materials, and is not related to the success of `GenCerts`. So, I think there is no need to use `t. Fatal`.实际上， `internetIP, err := utils.InternetIP()` 和 `internetIP, err := utils.InternetIP()` 只是生成一些非必要的材料，与 `GenCerts` 的成功无关。所以，我认为没有必要使用 `t. Fatal` 。

Ok, I think you're probably referring to hostName, err := os.Hostname(); err ! = nil and internetIP, err := utils.InternetIP()? How about I modify them to use klog.Infof to output a prompt?

zhzhuang-zju replied:
How about `klog.Warningf`?

tiansuo114 replied:
> How about `klog.Warningf`?

Ok, I modified the code according to your suggestion, how does it look now?

zhzhuang-zju replied:
thanks~ just a little typo problem. 
```klog.Warningf("Failed to get the current hostname,error message:%s. ", err)``` to ```klog.Warningf("Failed to get the current hostname, error message: %s.", err)``` 
```klog.Warningf("Failed to obtain internet IP,error message:%s. ", err)``` to ```klog.Warningf("Failed to obtain internet IP, error message: %s.", err)```
And this part looks good to me.

tiansuo114 replied:
Okay, so I updated the code and rebase it

tiansuo114 commented on pkg/karmadactl/cmdinit/cert/cert_test.go:
```diff
@@ -106,7 +123,68 @@ func TestGenCerts(t *testing.T) {
+		klog.Infof("All certificate files are present with CA certificates address parameter exists")
+	}
+
+	if ok, err := compareCertFilesInDirs(TestCertsTmp, TestCertsTmpWithArg, "ca.crt"); !ok || err != nil {
```
What should I do to solve the problem here? For example, I designed separate subfolders for two GenCerts in one folder, or I misunderstood your meaning

I consider adding the function of the comparison file to determine whether the ca certificate before and after is the same, maybe I have a design error here, please give further modification plan

liangyuanpeng replied:
You're right,  I misunderstood your meaning.  `compareCertFilesInDirs` is working for compare the files from two test case.

I think this is not necessary, WDYT  @chaosi-zju  @zhzhuang-zju

zhzhuang-zju replied:
refer to https://github.com/karmada-io/karmada/blob/a87ec2a61abadf9f102440b1f64cf338b2ee5518/pkg/karmadactl/cmdinit/cert/cert.go#L314-L333
When the ca exists, the existing ca is used, otherwise the ca is regenerated. I think it is necessary to compare the two ca files.
First，`GenCerts(TestCertsTmp, "", "", etcdServerCertConfig, etcdClientCertCfg, karmadaCertCfg, apiserverCertCfg, frontProxyClientCertCfg)` will generate files `./test-certs-tmp-without-ca-certificate/ca.crt` and `./test-certs-tmp-without-ca-certificate/ca.key`，and then `GenCerts(TestCertsTmpWithArg, TestCaCertPath, TestCaKeyPath, etcdServerCertConfig, etcdClientCertCfg, karmadaCertCfg, apiserverCertCfg, frontProxyClientCertCfg)` will use them instead of regenerating new ones. Comparing the two files verifies exactly this scenario

liangyuanpeng replied:
Thanks for your clarify.

/lgtm

tiansuo114 commented:
How is this pr now? Can it be incorporated? Or is there some discussion to be discussed? If there are more changes to be made, I will make timely changes

zhzhuang-zju commented:
> How is this pr now? Can it be incorporated? Or is there some discussion to be discussed? If there are more changes to be made, I will make timely changes

thanks~ @liangyuanpeng @chaosi-zju PTAL

chaosi-zju commented:
/lgtm

XiShanYongYe-Chang reviewed (commented):
Thanks~
/approve
